File: ChatAI/Java/Web/SpringBoot/src/main/java/com/portfolio/chatai/provider/ChatProviderFactory.java
package com.portfolio.chatai.provider;

import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.http.client.SimpleClientHttpRequestFactory;
import org.springframework.stereotype.Component;
import org.springframework.web.client.RestTemplate;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;

@Component
public class ChatProviderFactory {

    private final ObjectMapper mapper;

    public ChatProviderFactory(ObjectMapper mapper) {
        this.mapper = mapper;
    }

    public String resolve(String requested) {
        if (requested != null && !requested.isBlank()) {
            return requested;
        }
        return System.getenv().getOrDefault("CHAT_PROVIDER", "openai");
    }

    public String fallbackProvider() {
        String fb = System.getenv("CHAT_FALLBACK_PROVIDER");
        return (fb != null && !fb.isBlank()) ? fb : null;
    }

    public int timeoutMs() {
        try {
            return Integer.parseInt(System.getenv().getOrDefault("CHAT_TIMEOUT_MS", "30000"));
        } catch (NumberFormatException e) {
            return 30000;
        }
    }

    public boolean ragEnabled() {
        String raw = System.getenv("RAG_ENABLED");
        return raw != null && List.of("1", "true", "TRUE", "yes").contains(raw);
    }

    public String ragSearchUrl() {
        return System.getenv().getOrDefault("RAG_SEARCH_URL", "http://semantic-search:5000/api/search");
    }

    public int ragTopK() {
        try {
            return Integer.parseInt(System.getenv().getOrDefault("RAG_TOP_K", "3"));
        } catch (NumberFormatException e) {
            return 3;
        }
    }

    public List<String> retrieveContext(String query) {
        try {
            String url = ragSearchUrl() + "?q=" + URLEncoder.encode(query, StandardCharsets.UTF_8) + "&k=" + ragTopK();
            SimpleClientHttpRequestFactory httpFactory = new SimpleClientHttpRequestFactory();
            httpFactory.setConnectTimeout(timeoutMs());
            httpFactory.setReadTimeout(timeoutMs());
            RestTemplate restTemplate = new RestTemplate(httpFactory);
            Map<String, Object> data = restTemplate.getForObject(url, Map.class);
            if (data == null) {
                return List.of();
            }
            Object resultsObj = data.get("results");
            if (!(resultsObj instanceof List)) {
                return List.of();
            }
            List<String> documents = new ArrayList<>();
            for (Object result : (List<?>) resultsObj) {
                if (result instanceof Map<?, ?> resultMap) {
                    Object document = resultMap.get("document");
                    if (document instanceof String doc && !doc.isBlank()) {
                        documents.add(doc);
                    }
                }
            }
            return documents;
        } catch (Exception e) {
            return List.of();
        }
    }

    public String keyFor(String provider) {
        return switch (provider) {
            case "openai", "openai-compatible" -> System.getenv("OPENAI_API_KEY");
            case "azure" -> System.getenv("AZURE_OPENAI_API_KEY");
            case "google" -> System.getenv("GOOGLE_API_KEY");
            case "anthropic" -> System.getenv("ANTHROPIC_API_KEY");
            default -> null;
        };
    }

    public String baseUrlFor(String provider) {
        return switch (provider) {
            case "openai", "openai-compatible" ->
                    System.getenv().getOrDefault("OPENAI_BASE_URL", "https://api.openai.com/v1");
            case "azure" ->
                    System.getenv().getOrDefault("AZURE_OPENAI_ENDPOINT", "https://api.openai.com/v1");
            case "google" ->
                    System.getenv().getOrDefault("GOOGLE_BASE_URL", "https://generativelanguage.googleapis.com");
            case "anthropic" ->
                    System.getenv().getOrDefault("ANTHROPIC_BASE_URL", "https://api.anthropic.com");
            default -> throw new IllegalArgumentException("Unsupported provider: " + provider);
        };
    }

    public String deploymentFor(String provider) {
        return switch (provider) {
            case "azure" -> System.getenv().getOrDefault("AZURE_OPENAI_DEPLOYMENT", "gpt-4o-mini");
            default -> null;
        };
    }

    public IChatProvider create(String provider) {
        provider = (provider == null || provider.isBlank()) ? resolve(null) : provider;
        String key = keyFor(provider);
        if (key == null || key.isBlank()) {
            throw new ProviderNotConfiguredException(provider);
        }
        String url = baseUrlFor(provider);
        int timeoutMs = timeoutMs();
        return switch (provider) {
            case "openai", "openai-compatible" ->
                    new OpenAiCompatibleChatProvider(mapper, url, key, timeoutMs);
            case "azure" ->
                    new AzureChatProvider(mapper, url, key, deploymentFor(provider), timeoutMs);
            case "google" ->
                    new GoogleChatProvider(mapper, url, key, timeoutMs);
            case "anthropic" ->
                    new AnthropicChatProvider(mapper, url, key, timeoutMs);
            default -> throw new IllegalArgumentException("Unsupported provider: " + provider);
        };
    }
}

File: ChatAI/CSharp/Web/AspNetMinimalApi/src/Providers/ChatProviderFactory.cs
using System.Net.Http.Headers;
using System.Text.Json;

public class ProviderCallException : Exception
{
    public ProviderCallException(string message) : base(message) { }
}

public class ProviderNotConfiguredException : Exception
{
    public string Provider { get; }

    public ProviderNotConfiguredException(string provider)
        : base($"Provider '{provider}' is not configured (missing API key)")
    {
        Provider = provider;
    }
}

public class ChatProviderFactory
{
    private static readonly JsonSerializerOptions JsonOptions = new()
    {
        PropertyNameCaseInsensitive = true
    };

    private static readonly HttpClient RagClient = new();

    public virtual string Resolve(string? requested)
    {
        if (!string.IsNullOrWhiteSpace(requested))
            return requested;
        return Environment.GetEnvironmentVariable("CHAT_PROVIDER") ?? "openai";
    }

    public virtual string? FallbackProvider()
    {
        var fb = Environment.GetEnvironmentVariable("CHAT_FALLBACK_PROVIDER");
        return string.IsNullOrWhiteSpace(fb) ? null : fb;
    }

    public virtual int TimeoutMs()
    {
        var raw = Environment.GetEnvironmentVariable("CHAT_TIMEOUT_MS") ?? "30000";
        return int.TryParse(raw, out var v) ? v : 30000;
    }

    public virtual bool RagEnabled()
    {
        var raw = Environment.GetEnvironmentVariable("RAG_ENABLED");
        return raw is "1" or "true" or "TRUE" or "yes";
    }

    public virtual string RagSearchUrl() =>
        Env("RAG_SEARCH_URL", "http://semantic-search:5000/api/search");

    public virtual int RagTopK()
    {
        var raw = Environment.GetEnvironmentVariable("RAG_TOP_K") ?? "3";
        return int.TryParse(raw, out var v) ? v : 3;
    }

    public virtual async Task<List<string>?> RetrieveContextAsync(string query)
    {
        try
        {
            RagClient.Timeout = TimeSpan.FromMilliseconds(TimeoutMs());
            var url = $"{RagSearchUrl()}?q={Uri.EscapeDataString(query)}&k={RagTopK()}";
            var response = await RagClient.GetAsync(url);
            if (!response.IsSuccessStatusCode)
                return null;
            var body = await response.Content.ReadAsStringAsync();
            using var doc = JsonDocument.Parse(body);
            if (!doc.RootElement.TryGetProperty("results", out var results))
                return null;
            var documents = new List<string>();
            foreach (var item in results.EnumerateArray())
            {
                if (item.TryGetProperty("document", out var document) &&
                    !string.IsNullOrWhiteSpace(document.GetString()))
                    documents.Add(document.GetString()!);
            }
            return documents;
        }
        catch
        {
            return null;
        }
    }

    public virtual string? KeyFor(string provider) => provider switch
    {
        "openai" or "openai-compatible" => Environment.GetEnvironmentVariable("OPENAI_API_KEY"),
        "azure" => Environment.GetEnvironmentVariable("AZURE_OPENAI_API_KEY"),
        "google" => Environment.GetEnvironmentVariable("GOOGLE_API_KEY"),
        "anthropic" => Environment.GetEnvironmentVariable("ANTHROPIC_API_KEY"),
        _ => null,
    };

    public virtual string BaseUrlFor(string provider) => provider switch
    {
        "openai" or "openai-compatible" =>
            Env("OPENAI_BASE_URL", "https://api.openai.com/v1"),
        "azure" =>
            Env("AZURE_OPENAI_ENDPOINT", "https://api.openai.com/v1"),
        "google" =>
            Env("GOOGLE_BASE_URL", "https://generativelanguage.googleapis.com"),
        "anthropic" =>
            Env("ANTHROPIC_BASE_URL", "https://api.anthropic.com"),
        _ => throw new ArgumentException($"Unsupported provider: {provider}"),
    };

    public virtual string DeploymentFor(string provider) =>
        provider == "azure" ? Env("AZURE_OPENAI_DEPLOYMENT", "gpt-4o-mini") : null!;

    public virtual IChatProvider Create(string provider)
    {
        var key = KeyFor(provider);
        if (string.IsNullOrWhiteSpace(key))
            throw new ProviderNotConfiguredException(provider);

        var url = BaseUrlFor(provider);
        var timeoutMs = TimeoutMs();

        return provider switch
        {
            "openai" or "openai-compatible" => new OpenAiCompatibleChatProvider(url, key, timeoutMs),
            "azure" => new AzureChatProvider(url, key, DeploymentFor("azure")!, timeoutMs),
            "google" => new GoogleChatProvider(url, key, timeoutMs),
            "anthropic" => new AnthropicChatProvider(url, key, timeoutMs),
            _ => throw new ArgumentException($"Unsupported provider: {provider}"),
        };
    }

    private static string Env(string name, string fallback)
    {
        var v = Environment.GetEnvironmentVariable(name);
        return string.IsNullOrWhiteSpace(v) ? fallback : v;
    }
}
